### PR TITLE
Node transform cache

### PR DESCRIPTION
**Pull request by Maksims**
Change logic of how entities transformation is calculated and cached.

Previously to get transform, traversion from node to root and then sync from root to node back was required on every call that requires world transform. Like `getPosition` or `getRotation`, etc.
This would lead to huge number of `sync` calls which would do simple checks, but it would add up. In scenes with many entities, this would lead to thousands, sometimes tens thousands of sync calls.

Now the algorithm changes slightly. When transformation is modified, it will ensure that all descendants are dirtified. Consequent transformations will not re-dirtify. Due to this rule, if node is dirty, then all descendants should have `_dirtyWorld == true` as well. But not all ascendants should have `_dirtyWorld == true`.
This way we can limit traversion and know at the moment of getting transform if it needs to re-sync with hierarchy. And even in such case, not all the way to root, but only to earliest dirty node in a tree up towards root.

This reduces number of `sync` calls from 5 to 1000 times, and in more complex scenes even more.
Small extra price of transforming unchanged nodes with many descendants, but much cheaper retrieval of transformation, especially if nodes are "static".

No external code should modify private dirty flags, but can call `_dirtify` method if very-very needed.

Additional bug fix with `event.fire`, where in rare cases of some modification to same event during callback would lead to an exception.

### CODE REVIEW

File: src/scene/forward-renderer.js [hunk 1]
```diff
@@ -1511,10 +1511,10 @@ pc.extend(pc, function () {
 
                 if (normal) {
                     normalMatrix = meshInstance.node.normalMatrix;
-                    if (meshInstance.node.dirtyNormal) {
+                    if (meshInstance.node._dirtyNormals) {
```
Comment (willeastcott) [inline src/scene/forward-renderer.js]: This should be _dirtyNormal. Remove the 's'.

**Comment by Maksims** [comment]
Worth noting, that there is another possible optimization to minimize number of syncHierarchy calls by marking nodes all the way to the root from dirtified node with some extra flag.
This would lead to almost no extra cost for static entities in hierarchy as they would be never even checked, unless there was transformations on the path to them.

**Comment by vkalpias** [comment]
I think we should make sure that aabbVer wraps correctly when it reaches NUMBER.MAX_VALUE because we are just increasing it and it will not wrap automatically.

**Comment by willeastcott** [comment]
I guarantee that no one will ever see aabbVer reach max value. :smile:

**Comment by vkalpias** [comment]
Really? What if something moves every frame

**Comment by willeastcott** [comment]
Well, do the math, but MAX_VALUE is truly enormous.

**Comment by vkalpias** [comment]
Hm I thought max_value was smaller

**Comment by guycalledfrank** [comment]
I'll wrap.

**Pull request merged**